src/node/mod.rs

```diff
@@ -12,6 +12,76 @@
 // If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 //
 
+//! Simplicity Program Nodes
+//!
+//! The types in this module are used to represent Simplicity expressions as
+//! DAGs. The nodes of this DAG are individual combinators along with some
+//! cached data about the node, which depends on the specific node type.
+//!
+//! All nodes represent the root of an expression. Expressions whose source
+//! and target types are both unit are called "programs". Generally speaking,
+//! any nodes that users hold directly will be programs.
+//!
+//! There are three main node types:
+//!
+//! 1. [`ReedemNode`] represents a Simplicity node as it exists on the blockchain.
+//!    Every witness node is populated with a correctly-typed [`Value`], every
+//!    disconnect node has a child, and every non-executed branch of a `case`
+//!    combinator is replaced by a hidden node, and there is nothing you can do
+//!    to modify the expression.
+//!
+//!    `RedeemNode`s can be executed on the bit machine and have complete types
+//!    and resource bounds, and can be serialized in the consensus bit-encoding.
+//!
+//! 2. [`CommitNode`] represents a Simplicity node as it is *committed to* on
+//!    the blockchain. This means that witness and (TODO) disconnect nodes are not
+//!    populated, but type inference is complete and resource bounds are
+//!    available. `case` combinators may have hidden children, or not.
+//!
+//!    There is a bit-encoding for `CommitNode`s which is essentially only used
+//!    by this library. It consists of the bit-encoding.of the combinators, fully
+//!    shared *except* that witness and disconnect nodes (and their ancestors)
+//!    are unshared. No witness data is included.
+//!
+//!    TODO there is also a human-readable encoding.
+//!
+//! 3. [`ConstructNode`] represents an "under-construction" Simplicity expression.
+//!    These nodes' types are not necessarily complete, and are inferred as the
+//!    program is constructed. This is the only node that you can programmatically
+//!    construct. It has no encoding, human-readable or bitwise, and is intended
+//!    to exist only ephemerally.
+//!
+//! The following conversions are possible between the node types:
+//!
+//! 1. [`ConstructNode::finalize_types`] converts a [`ConstructNode`] to a
+//!    [`CommitNode`] by setting any free variables, as well as the source and
+//!    target of the root node, to unit.
+//!
+//!    This conversion requires no input from the user but may fail with a type
+//!    error, in case of infinitely-sized types or in the case that the unit
+//!    bounds cannot be applied.
+//!
+//! 2. [`CommitNode::rtl_finalize`] converts a [`CommitNode`] to a [`RedeemNode`]
```

Comment:
bebad27e1959bbd1f6e7229676035e139ae82fb7: `CommitNode::rtl_finalize` does not exist in the final diff. Also, what does `rtl` mean?

Reply:
Yes, it will come in a later PR. Would you like me to delete this comment for now? `rtl` stands for `right-to-left`.